README
NAME
    HTTP::MobileAgent - HTTP mobile user agent string parser

SYNOPSIS
      use HTTP::MobileAgent;

      # from PSGI $env hash:
      my $agent = HTTP::MobileAgent->new( $env );

      # from %ENV (CGI mode):
      my $agent = HTTP::MobileAgent->new();

      # from Apache (mod_perl 1.x):
      my $agent = HTTP::MobileAgent->new(Apache->request);

      # from a HTTP::Headers / HTTP::Request object:
      my $agent = HTTP::MobileAgent->new( HTTP::Headers->new( ... ) );
      my $agent = HTTP::MobileAgent->new( HTTP::Request->new( ... ) );

      # from a raw user agent string:
      my $agent = HTTP::MobileAgent->new($agent_string);

      if ($agent->is_docomo) {
          # or if ($agent->name eq 'DoCoMo')
          # or if ($agent->isa('HTTP::MobileAgent::DoCoMo'))
          # it's NTT DoCoMo i-mode.
          # see what's available in H::MA::DoCoMo
      } elsif ($agent->is_vodafone) {
          # it's Vodafone(J-Phone).
          # see what's available in H::MA::Vodafone
      } elsif ($agent->is_ezweb) {
          # it's KDDI/EZWeb.
          # see what's available in H::MA::EZweb
      } else {
          # may be PC
          # $agent is H::MA::NonMobile
      }

      my $display = $agent->display;        # HTTP::MobileAgent::Display
      if ($display->color) { ... }

DESCRIPTION
    HTTP::MobileAgent parses HTTP_USER_AGENT strings of (mainly Japanese)
    mobile HTTP user agents. It'll be useful in page dispatching by user
    agents.

METHODS
    Here are common methods of HTTP::MobileAgent subclasses. More agent
    specific methods are described in each subclasses. Note that some of
    common methods are also overrided in some subclasses.

    new
          $agent = HTTP::MobileAgent->new;       # from %ENV
          $agent = HTTP::MobileAgent->new($env); # PSGI env hash
          $agent = HTTP::MobileAgent->new($r);   # Apache or HTTP::Request
          $agent = HTTP::MobileAgent->new($ua_string);

        Parses HTTP headers and constructs HTTP::MobileAgent subclass
        instance.

        If no argument is supplied, $ENV{HTTP_*} is used (i.e., expects a
        CGI environment to be setup)

        If a single hash reference is given, then that hash is treated as a
        PSGI environment hash.

        If a blessed reference which is based on Apache (mod_perl 1.x),
        HTTP::Headers or HTTP::Request is passed, those will be used
        accordingly to parse data.

        If a single scalar is given, then that is taken to be a raw user
        agent string. Note that most likely this form of usage will not give
        you much information, as some mobile agents put useful information
        on HTTP headers other than only "User-Agent:" (like
        "x-jphone-msname" in J-Phone).

    user_agent
          print "User-Agent: ", $agent->user_agent;

        returns User-Agent string.

    name
          print "name: ", $agent->name;

        returns User-Agent name like 'DoCoMo'.

    is_docomo, is_vodafone(is_j_phone, is_softbank), is_ezweb, is_wap1,
    is_wap2, is_tuka,is_non_mobile
           if ($agent->is_docomo) { }

        returns if the agent is DoCoMo, Vodafone(J-Phone) or EZweb.

    carrier
          print "carrier: ", $agent->carrier;

    carrier_longname
          print "carrier_longname: ", $agent->carrier_longname;

    display
          my $display = $agent->display;

        returns HTTP::MobileAgent::Display object. See
        HTTP::MobileAgent::Display for details.

    user_id
          my $user_id = $agent->user_id;

        return X-DCMGUID, X-UP-SUBNO or X-JPHONE-UID.

    gps_compliant
          if ($agent->gps_compliant) { }

        returns if the agent is GPS compliant.

NOTE
    "Why not adding this module as an extension of HTTP::BrowserDetect?"
        Yep, I tried to do. But the module's code seems hard enough for me
        to extend and don't want to bother the author for this
        mobile-specific features. So I made this module as a separated one.

MORE IMPLEMENTATIONS
    If you have any idea / request for this module to add new subclass, I'm
    open to the discussion or (more preferable) patches. Feel free to mail
    me.

OTHER LANGUAGE BINDINGS
    This module is now ported to PHP as Net::UserAgent::Mobile by Atsuhiro
    KUBO. See http://pear.php.net/package-info.php?pacid=180 for details.

AUTHOR
    Tatsuhiko Miyagawa <miyagawa@bulknews.net> is the original author and
    wrote almost all the code.

    with contributions of Satoshi Tanimoto <tanimoto@cpan.org> and Yoshiki
    Kurihara <kurihara@cpan.org>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

MAIN
SEE ALSO
    HTTP::MobileAgent::DoCoMo, HTTP::MobileAgent::Vodafone,
    HTTP::MobileAgent::JPhone, HTTP::MobileAgent::EZweb,
    HTTP::MobileAgent::NonMobile, HTTP::MobileAgent::Display,
    HTTP::BrowserDetect

    Reference URL for specification is listed in Pods for each subclass.



Changes
Revision history for Perl extension HTTP::MobileAgent

0.36 Tue Jul 24 12:38:16 2012
     * Fixed mismatch between some models and ua
       (Thanks to matuu).
     * Added new x-up-devcap-multimedia(StandAloneGPS) in EZweb.pm
       (Thanks to tokuhirom).
     * Added CA01C F02D F04D F06D F07C F08A3 F09A3 F09C F10C F11C L10C
       N02D N03D N05C N06A3 N07A3 N08A3 N09A3 P03D P04C P05C P06C P07A3
       P08A3 P09A3 SH03D SH05A3 SH05D SH06A3 SH07A3 SH10C SH11C
       in DoCoMoDisplayMap and DoCoMo.xml (auto-generated)
     * Removed D209I D210I D211I D251I D251IS D252I D253I D253IWM D501I
       D502I D503I D503IS D504I D505I D505IS D506I ER209I F08A F09A F209I
       F210I F211I F212I F251I F501I F502I F502IT F503I F503IS F504I F504IS
       F505I F505IGPS F506I F661I F671I F671IS F672I KO209I KO210I N06A N07A
       N08A N09A N209I N210I N211I N211IS N251I N251IS N252I N253I N501I
       N502I N502IT N503I N503IS N504I N504IS N505I N505IS N506I N506IS
       N506ISII N821I NM502I P07A P08A P09A P209I P209IS P210I P211I P211IS
       P213I P251IS P252I P252IS P253I P253IS P501I P502I P503I P503IS
       P504I P504IS P505I P505IS P506IC P506ICII P651PS P821I R209I R211I
       R691I R692I SH05A SH06A SH07A SH251I SH251IS SH252I SH505I SH505IS
       SH506IC SH821I SO210I SO211I SO212I SO213I SO213IS SO213IWR SO502I
       SO502IWM SO503I SO503IS SO504I SO505I SO505IS SO506I SO506IC SO506IS
       from DoCoMoDisplayMap and DoCoMo.xml (auto-generated)

0.35 Tue Jan 10 20:50:36 2012
    * Updated $HTMLVerMap and $GPSModelsRe in DoCoMo.pm.

0.33 Tue Apr 26 18:18:49 2011
    * Fixed condition of is_win() for new some ezweb terminals
      (ex. KC41, KC42, KC43) in EZweb.pm.
    - Added F08C SH08C SH09C in DoCoMoDisplayMap (auto-generated)
    - Added F08C SH08C SH09C in DoCoMoMap.xml (auto-generated)

0.32 Thu Feb 24 12:17:49 2011
    * Added _make_display(), user_id(), gps_compliant() in NonMobile.pm.
      (Thanks to tokuhirom)
    - Added F05C SH06C in DoCoMoDisplayMap (auto-generated)
    - Added F05C SH06C in DoCoMoMap.xml (auto-generated)

0.31 Fri Feb  4 14:20:07 2011
    * Work with PSGI env hash (lestrrat)

0.30 Fri Jan 28 12:23:52 2011
    * Fixed regexp to match with Googlebot-Mobile/2.1
      in DoCoMo.pm(Thanks to ziguzagu).
      https://rt.cpan.org/Ticket/Display.html?id=50700
    * Changed $GPSModelsRe to disable SH905iTV in DoCoMo.pm(Thanks to hirafoo).
    * Fixed gps_compliant() how to get 'x-up-devcap-multimedia'
      in EZweb.pm(Thanks to nekokak).
    - Added F01C F02C F03C F04C L01C L03C N01C N02C N03C P01C P02C
      P03C SH01C SH02C SH04C SH05C in DoCoMoDisplayMap (auto-generated)
    - Added F01C F02C F03C F04C L01C L03C N01C N02C N03C P01C P02C
      P03C SH01C SH02C SH04C SH05C in DoCoMoMap.xml (auto-generated)

0.29 Tue Sep  7 05:42:45 2010
    * Added gps_compliant() in AirHPhone.pm
      for HTTP::MobileAgent::Plugin::Locator compatibility.
    * Removed $GPSModels in DoCoMo.pm.
    * Updated $HTMLVerMap and $GPSModelsRe in DoCoMo.pm.
    * Added diff in makexmldocomomap.
    - Added F01B F02B F03B F04B F06B F07B F08B F09B F10A F10B L01B N01B
      N02B N03B N04B N05B N06B N07A N07B N08A N08B N09A P01B P02B P03B
      P04B P05B P06B P07B P08A P09A SH01B SH02B SH03B SH04B SH05B SH06A
      SH06B SH07A SH07B SH08A SH08B SH09B in DoCoMoDisplayMap (auto-generated)
    - Added F01B F02B F03B F04B F06B F07B F08B F09B F10A F10B L01B N01B
      N02B N03B N04B N05B N06B N07A N07B N08A N08B N09A P01B P02B P03B
      P04B P05B P06B P07B P08A P09A SH01B SH02B SH03B SH04B SH05B SH06A
      SH06B SH07A SH07B SH08A SH08B SH09B in DoCoMoMap.xml (auto-generated)

0.28 Tue Jul  7 16:02:33 2009
    * Using HTTP::Headers::Fast (Thanks to Tokuhiro MATSUNO)
    * Added gps_compliant() method in DoCoMo.pm/EZweb.pm/Vodafone.pm(SoftBank).
    * Added browser_version() method in DoCoMo.pm.
    * $GPSModels and is_gps() method was obsolete
      but leaved for backward compatibility in DoCoMo.pm.
    * Fixed user_id() problem in Vodafone.pm(SoftBank). (Thanks to masutaro)
    * Updated $HTMLVerMap.
    - Added NM705I NM706I SH01A SH02A SH03A SH04A SH706I SH706IW SH905ITV
      SH906I SH906ITV F01A F03A F06A F905IBIZ F906I SO905ICS SO906I SH05A
      F09A D705IMYU F07A L705I L705IX L706IE F884IES D705I N01A N02A N03A
      N04A N05A N705I N705IMYU N706I N706IE N706III N905IBIZ N906I N906IL
      N906IMYU SH705I SH705III SH706IE SO705I SO706I P01A P02A P03A P04A
      P05A P06A P10A P705I P705ICL P705IMYU P706IE P706IMYU P905ITV P906I
      F02A F04A F705I F706I F884I N06A F08A P07A F883IESS F883IS L852I F05A F801I
      in DoCoMoMap.xml (auto-generated)
    - Added F01A F02A F03A F04A F05A F06A F07A F08A F09A F706I
      F883IESS F883IS F884I F884IES F906I L706IE L852I
      N01A N02A N03A N04A N05A N06A N706I N706IE N706III N906I N906IL N906IMYU
      NM706I P01A P02A P03A P04A P05A P06A P07A P10A P706IE P706IMYU P906I
      SH01A SH02A SH03A SH04A SH05A SH706I SH706IE SH706IW SH906I SH906ITV
      SO706I SO906I in DoCoMoDisplayMap (auto-generated)

0.27 Fri Apr 11 13:34:23 2008
    * Supported user_id to get guid in DoCoMo
    * Using WWW::MobileCarrierJP when get terminal display information
      (Thanks to Tokuhiro MATSUNO)

0.26_01 Thu Aug  9 17:08:05 2007
    * Supported funny mobile search crawler.
      eg. J-Phone/5.0/J-SH03
          Nokia6820/2.0 (4.83) Profile/MIDP-1.0 Configuration/CLDC-1.0
      (Thanks to Akira Kamikura and Tokuhiro MATSUNO)
    - Added L600I L601I M702IS M702IG L602I F902IS SO902IWP+
      N902IX N902IL N702IS P702ID D702IF F882IES N601I D800IDS
      P703I F883I P704I SH903I P903I N903I D903I F903I SO903I
      D903ITV F903IX P903ITV SH903ITV F903IBSC P903IX SO903ITV
      N703ID F703I P703I D703I SH703I N703I SO703I SH904I N904I
      F904I D904I P904I SO704I F704I N704I SH704I D704I F883IES
      in DoCoMoMap.xml (auto-generated)
    - Added D903ITV SO903ITV F704I SO703I L602I D800IDS D703I D702IF

lib/HTTP/MobileAgent.pm
package HTTP::MobileAgent;

use strict;
use vars qw($VERSION);
$VERSION = '0.36';

use HTTP::MobileAgent::Request;

require HTTP::MobileAgent::DoCoMo;
require HTTP::MobileAgent::JPhone;
require HTTP::MobileAgent::EZweb;
require HTTP::MobileAgent::AirHPhone;
require HTTP::MobileAgent::NonMobile;
require HTTP::MobileAgent::Display;

use vars qw($MobileAgentRE);
# this matching should be robust enough
# detailed analysis is done in subclass's parse()
my $DoCoMoRE = '^DoCoMo/\d\.\d[ /]';
my $JPhoneRE = '^(?i:J-PHONE/\d\.\d)';
my $VodafoneRE = '^Vodafone/\d\.\d';
my $VodafoneMotRE = '^MOT-';
my $SoftBankRE = '^SoftBank/\d\.\d';
my $SoftBankCrawlerRE = '^Nokia[^/]+/\d\.\d';
my $EZwebRE  = '^(?:KDDI-[A-Z]+\d+[A-Z]? )?UP\.Browser\/';
my $AirHRE = '^Mozilla/3\.0\((?:WILLCOM|DDIPOCKET)\;';
$MobileAgentRE = qr/(?:($DoCoMoRE)|($JPhoneRE|$VodafoneRE|$VodafoneMotRE|$SoftBankRE|$SoftBankCrawlerRE)|($EZwebRE)|($AirHRE))/;

sub new {
    my($class, $stuff) = @_;
    my $request = HTTP::MobileAgent::Request->new($stuff);

    # parse UA string
    my $ua = $request->get('User-Agent');
    my $sub = 'NonMobile';
    if ($ua =~ /$MobileAgentRE/) {
        $sub = $1 ? 'DoCoMo' : $2 ? 'JPhone' : $3 ? 'EZweb' :  'AirHPhone';
    }

    my $self = bless { _request => $request }, "$class\::$sub";
    $self->parse;
    return $self;
}


sub user_agent {
    my $self = shift;
    $self->get_header('User-Agent');
}

sub get_header {
    my($self, $header) = @_;
    $self->{_request}->get($header);
}

# should be implemented in subclasses
sub parse { die }
sub _make_display { die }

sub name  { shift->{name} }

sub display {
    my $self = shift;
    unless ($self->{display}) {
	$self->{display} = $self->_make_display;
    }
    return $self->{display};
}

# utility for subclasses
sub make_accessors {
    my($class, @attr) = @_;
    for my $attr (@attr) {
	no strict 'refs';
	*{"$class\::$attr"} = sub { shift->{$attr} };
    }
}

sub no_match {
    my $self = shift;
    require Carp;
    Carp::carp($self->user_agent, ": no match. Might be new variants. ",
	       "please contact the author of HTTP::MobileAgent!") if $^W;
}

sub is_docomo  { 0 }
sub is_j_phone { 0 }
sub is_vodafone { 0 }
sub is_softbank { 0 }
sub is_ezweb   { 0 }
sub is_airh_phone { 0 }
sub is_non_mobile { 0 }
sub is_tuka { 0 }

sub is_wap1 {
    my $self = shift;
    $self->is_ezweb && ! $self->is_wap2;
}

sub is_wap2 {
    my $self = shift;
    $self->is_ezweb && $self->xhtml_compliant;
}

sub carrier { undef }
sub carrier_longname { undef }

1;
__END__

=head1 NAME

HTTP::MobileAgent - HTTP mobile user agent string parser

=head1 SYNOPSIS

  use HTTP::MobileAgent;

  # from PSGI $env hash:
  my $agent = HTTP::MobileAgent->new( $env );

  # from %ENV (CGI mode):
  my $agent = HTTP::MobileAgent->new();

  # from Apache (mod_perl 1.x):
  my $agent = HTTP::MobileAgent->new(Apache->request);

  # from a HTTP::Headers / HTTP::Request object:
  my $agent = HTTP::MobileAgent->new( HTTP::Headers->new( ... ) );
  my $agent = HTTP::MobileAgent->new( HTTP::Request->new( ... ) );

  # from a raw user agent string:
  my $agent = HTTP::MobileAgent->new($agent_string);

  if ($agent->is_docomo) {
      # or if ($agent->name eq 'DoCoMo')
      # or if ($agent->isa('HTTP::MobileAgent::DoCoMo'))
      # it's NTT DoCoMo i-mode.
      # see what's available in H::MA::DoCoMo
  } elsif ($agent->is_vodafone) {
      # it's Vodafone(J-Phone).
      # see what's available in H::MA::Vodafone
  } elsif ($agent->is_ezweb) {
      # it's KDDI/EZWeb.
      # see what's available in H::MA::EZweb
  } else {
      # may be PC
      # $agent is H::MA::NonMobile
  }

  my $display = $agent->display;	# HTTP::MobileAgent::Display
  if ($display->color) { ... }

=head1 DESCRIPTION

HTTP::MobileAgent parses HTTP_USER_AGENT strings of (mainly Japanese)
mobile HTTP user agents. It'll be useful in page dispatching by user agents.

=head1 METHODS

Here are common methods of HTTP::MobileAgent subclasses. More agent
specific methods are described in each subclasses. Note that some of
common methods are also overrided in some subclasses.

=over 4

=item new

  $agent = HTTP::MobileAgent->new;       # from %ENV
  $agent = HTTP::MobileAgent->new($env); # PSGI env hash
  $agent = HTTP::MobileAgent->new($r);	 # Apache or HTTP::Request
  $agent = HTTP::MobileAgent->new($ua_string);

Parses HTTP headers and constructs HTTP::MobileAgent subclass
instance.

If no argument is supplied, $ENV{HTTP_*} is used (i.e., expects a CGI
environment to be setup)

If a single hash reference is given, then that hash is treated as a
PSGI environment hash.

If a blessed reference which is based on Apache (mod_perl 1.x),
HTTP::Headers or HTTP::Request is passed, those will be used 
accordingly to parse data.

If a single scalar is given, then that is taken to be a raw user agent
string. Note that most likely this form of usage will not give you much
information, as some mobile agents put useful information on HTTP headers
other than only C<User-Agent:> (like C<x-jphone-msname> in J-Phone).

=item user_agent

  print "User-Agent: ", $agent->user_agent;

returns User-Agent string.

=item name

  print "name: ", $agent->name;

returns User-Agent name like 'DoCoMo'.

=item is_docomo, is_vodafone(is_j_phone, is_softbank), is_ezweb, is_wap1, is_wap2, is_tuka,is_non_mobile

   if ($agent->is_docomo) { }

returns if the agent is DoCoMo, Vodafone(J-Phone) or EZweb.

=item carrier

  print "carrier: ", $agent->carrier;

=item carrier_longname

  print "carrier_longname: ", $agent->carrier_longname;

=item display

  my $display = $agent->display;

returns HTTP::MobileAgent::Display object. See
L<HTTP::MobileAgent::Display> for details.

=item user_id

  my $user_id = $agent->user_id;

return X-DCMGUID, X-UP-SUBNO or X-JPHONE-UID.

=item gps_compliant

  if ($agent->gps_compliant) { }

returns if the agent is GPS compliant.

=back

=head1 WARNINGS

Following warnings might be raised when C<$^W> is on.

=over 4

=item "%s: no match. Might be new variants. please contact the author of HTTP::MobileAgent!"

User-Agent: string does not match patterns provided in subclasses. It
may be faked user-agent or a new variant. Feel free to mail me to
inform this.

=back

=head1 NOTE

=over 4

=item "Why not adding this module as an extension of HTTP::BrowserDetect?"

Yep, I tried to do. But the module's code seems hard enough for me to
extend and don't want to bother the author for this mobile-specific
features. So I made this module as a separated one.

=back

=head1 MORE IMPLEMENTATIONS

If you have any idea / request for this module to add new subclass,
I'm open to the discussion or (more preferable) patches. Feel free to
mail me.

=head1 OTHER LANGUAGE BINDINGS

This module is now ported to PHP as Net::UserAgent::Mobile by Atsuhiro
KUBO.  See http://pear.php.net/package-info.php?pacid=180 for details.

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt> is the original author and wrote almost all the code.

with contributions of Satoshi Tanimoto E<lt>tanimoto@cpan.orgE<gt> and Yoshiki Kurihara E<lt>kurihara@cpan.orgE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 MAIN

=head1 SEE ALSO

L<HTTP::MobileAgent::DoCoMo>, L<HTTP::MobileAgent::Vodafone>, L<HTTP::MobileAgent::JPhone>,
L<HTTP::MobileAgent::EZweb>, L<HTTP::MobileAgent::NonMobile>,
L<HTTP::MobileAgent::Display>, L<HTTP::BrowserDetect>

Reference URL for specification is listed in Pods for each subclass.

=cut
